sdcm/cluster.py

```diff
@@ -2149,39 +2144,53 @@ def is_scylla_installed(self, raise_if_not_installed=False):
+    def get_scylla_binary_version(self) -> Optional[str]:
+        result = self.remoter.run(f"{self.add_install_prefix('/usr/bin/scylla')} --version", ignore_status=True)
         if result.ok:
-            self.scylla_version_detailed = result.stdout.strip()
-            if build_id := self.get_scylla_build_id():
-                self.scylla_version = self.scylla_version_detailed
-                self.scylla_version_detailed += f" with build-id {build_id}"
-                self.log.info(f"Found ScyllaDB version with details: {self.scylla_version_detailed}")
-            self.log.debug(f"self.scylla_version_detailed={self.scylla_version_detailed}")
-        else:
-            if self.distro.is_rhel_like:
-                cmd = f"rpm --query --queryformat '%{{VERSION}}' {self.scylla_pkg()}"
-            else:
-                cmd = f"dpkg-query --show --showformat '${{Version}}' {self.scylla_pkg()}"
-            result = self.remoter.run(cmd, ignore_status=True)
+            return result.stdout.strip()
+        self.log.debug("Unable to get ScyllaDB version using `%s':\n%s\n%s",
+                       result.command, result.stdout, result.stderr)
+        return None
 
-        if match := result.ok and SCYLLA_VERSION_RE.match(result.stdout):
-            self.scylla_version = match.group().replace("~", ".")
-            self.log.info("Found ScyllaDB version: %s", self.scylla_version)
+    def get_scylla_package_version(self) -> Optional[str]:
+        if self.distro.is_rhel_like:
+            cmd = f"rpm --query --queryformat '%{{VERSION}}' {self.scylla_pkg()}"
         else:
-            self.log.debug("Unable to get or parse ScyllaDB version using `%s':\n%s\n%s",
-                           cmd, result.stdout, result.stderr)
-
-        if not self.scylla_version:
-            self.log.warning(
-                "All attempts to get ScyllaDB version failed. Looks like there is no ScyllaDB installed.")
-
-        if not self.scylla_version_detailed:
-            self.scylla_version_detailed = self.scylla_version
+            cmd = f"dpkg-query --show --showformat '${{Version}}' {self.scylla_pkg()}"
+        result = self.remoter.run(cmd, ignore_status=True)
+        if result.ok:
+            return result.stdout.strip()
+        self.log.debug("Unable to get ScyllaDB version using `%s':\n%s\n%s",
+                       result.command, result.stdout, result.stderr)
+        return None
 
-        return self.scylla_version_detailed
+    @optional_cached_property
+    def scylla_version_detailed(self) -> Optional[str]:
+        scylla_version = self.get_scylla_binary_version()
+        if scylla_version is None:
+            if scylla_version := self.get_scylla_package_version():
+                return scylla_version.replace("~", ".")
+            self.log.warning("All attempts to get ScyllaDB version failed. Looks like there is no ScyllaDB installed.")
+            raise NoValue
```

Comment:
I think it would be useful to place the error message in the exception.

Reply:
This exception used as an auxiliary mechanism and not exposed anywhere outside of `optional_cached_property` stuff. How do you think to use this error message?

Reply:
Why then we cannot use builtin "ValueError"?

Reply:
In this case any ValueError thrown from the property function will trigger same behavior, which is not ok, we would want to see unexpected failure as exception, not as None on the output.

Reply:
Nothing to add to Dmitry's comment :)